{% extends 'base.html' %}
{% from "help.html" import help with context %}

{% block title %} {{ protein.gene.name }} - {{ protein.refseq }} - protein view - {{ super() }} {% endblock %}

{% block head %}
  {{ super() }}

  {% assets "css_protein" %}
    <link rel="stylesheet" href="{{ ASSET_URL }}">
  {% endassets %}

  {{ dependency('d3.js') }}
  {{ dependency('md5') }}

  {{ dependency('bootstrap_table') }}
  {{ dependency('bootstrap_table_css') }}

  {{ dependency('table_export') }}
  {{ dependency('bootstrap_table_export') }}

  {# Nunjucks templates #}
  {% if is_debug_mode %}
    {{ dependency('nunjucks') }}
  {% else %}
    {{ dependency('nunjucks_slim') }}
    <script type="text/javascript" src="{{ url_for('static', filename='/js_templates/precompiled/protein.js') }}"></script>
  {% endif %}

  {{ dependency('clipboard.js') }}

{% endblock %}


{% block breadcrumb %}
  {{ super() }}
  <li><a href="{{ url_for('ProteinView:index') }}">Proteins</a></li>
  <li><a href="{{ url_for('GeneView:show', gene_name=protein.gene.name) }}">{{ protein.gene.name }}</a></li>
  <li class="active">{{ protein.refseq }}</li>
{% endblock %}


{% block content %}

  <h3>{{ protein.gene.name }} ({{ protein.refseq }}) {% if protein.gene.full_name %} - {{ protein.gene.full_name }} {% endif %}</h3>
  <p class="full-protein-name">{% if protein.full_name %}{{ protein.full_name | title }}{% endif %}</p>

  <div class="isoform-page-options">
    {% with active_tab = 'sequence' %}
      {% include 'isoform_tabs_menu.html' %}
    {% endwith %}
    <div class="actions">
      {% include 'buttons/short_url.html' %}
    </div>
  </div>


  <div class="row">
    <div class="col-md-3">

      <div id="sticky_box">

        {# Filters #}
        <form method="GET" class="widget-form" id="filters_form">
          {% include 'filters/protein.html' with context %}
          <input type="hidden" name="fallback" value="True">
        </form>

        {# Legend #}
        <div class="panel panel-default color-legend">
          <div class="panel-heading">
            <h3 class="panel-title">Legend</h3>
          </div>
          <div class="panel-body">
            <div class="legend">
              <h4>
                Mutation impacts
                {{ help('legend-mutation-impacts', 'small-top') }}
              </h4>
              {% for type in mutation_types %}
                <div class="swatch">
                  <div class="impact {{ type }}">
                    <div class="color"></div>
                  </div>
                  <div class="name">{{ type }}</div>
                </div>
              {% endfor %}
            </div>
            <div class="legend">
              <h4>
                Sites
                {{ help('legend-sequence-sites', 'small-top') }}
              </h4>
              {% for type in site_types %}
                <div class="swatch">
                  <div class="site {{ type }}">
                    <div class="color"></div>
                  </div>
                  {% if type == 'multi_ptm' %}
                    <div class="name">multiple types</div>
                  {% else %}
                    <div class="name">{{ type }}</div>
                  {% endif %}
                </div>
              {% endfor %}
            </div>
            <div class="legend">
              <h4>Others</h4>
              <div class="swatch">
                <div class="site-position"></div>
                <div class="name long">Exact position of a PTM site</div>
              </div>
            </div>
          </div>
        </div>

        {# Contact us #}
        {% with feature='protein', title=protein.gene.name + ' - ' + protein.refseq %}
          {% include "contact_box.html" %}
        {% endwith %}

      </div>

    </div>
    <div class="col-md-9">


      {# Plot #}
      <div class="panel panel-default panel-needleplot panel-with-btn visualisation">
        <div class="panel-heading">
          <div class="heading-left">
            <h3 class="panel-title">
              Mutations Visualisation
              {{ help('mutations-visualisation') }}
            </h3>
          </div>
          <div class="heading-right">
            {% include 'buttons/export.html' %}
          </div>
        </div>
        <div class="panel-body">

          <div class="tracks-controls invisible">

            <div class="input-group">
              <div class="input-group-btn">
                <button type="button" class="btn btn-default scroll-left">
                  <span class="glyphicon glyphicon-chevron-left"></span>
                </button>
                <button type="button" class="btn btn-default scroll-right">
                  <span class="glyphicon glyphicon-chevron-right"></span>
                </button>
                <button type="button" class="btn btn-default zoom-in">
                  <span class="glyphicon glyphicon-zoom-in"></span>
                </button>
                <button type="button" class="btn btn-default zoom-out">
                  <span class="glyphicon glyphicon-zoom-out"></span>
                </button>
              </div>
              <input class="form-control scroll-to-input" placeholder="Type a position, e.g. 105 or 100-200">
              <span class="form-control-feedback scroll-feedback"></span>
              <span class="input-group-addon">amino acids</span>
              <span class="input-group-btn">
              <button class="btn btn-default scroll-to" type="button">Go!</button>
            </span>
            </div><!-- /input-group -->

          </div>

          <div id="protein_spinner" class="visualisation_spinner">
            <span class="glyphicon glyphicon-refresh glyphicon-spin"></span>
          </div>
          <div id="plot_area" class="transparent"></div>
        </div>
      </div>

      {# tracks #}
      <div class="panel panel-default panel-tracks">
        <div class="panel-body">
          <div id="tracks-box">
            {#  To be loaded with JS #}
            {#
            {% with tracks=tracks %}
              {% include 'protein/tracks.html' %}
            {% endwith %}
            #}
          </div>
        </div>
      </div>

      {# Summary #}
      {% include 'protein/summary.html' %}

      {# table #}
      <div class="table-outer-wrapper">
        <div class="table-help-btn hidden">
          {{ help('mutations-table-help-btn') }}
        </div>
        <div id="table-wrapper">
          {# The table will be filled with HTML over AJAX request for mutations details #}
        </div>
      </div>

      {# references #}
      {% include 'protein/external_references.html' %}

    </div>
  </div>

{% endblock %}


{% block footer_js %}
  {% assets "js_protein_view" %}
    <script type="text/javascript" src="{{ ASSET_URL }}"></script>
  {% endassets %}
  <script type="text/javascript">

    nunjucks.configure("{{ url_for('static', filename='js_templates') }}", {autoescape: true, web: {useCache: true}});
    nunjucks.installJinjaCompat()

    affix($('#sticky_box'), $('footer'));

    short_url = ShortURL();
    short_url.init(
      "{{ url_for('ShortAddress:get_shorthand_for') }}",
      "{{ url_for('ShortAddress:visit_shorthand', shorthand='<shorthand>', _external=True) }}"
    );

    var tracks = Tracks();

    var tracks_box = $('#tracks-box');
    var needle_plot = null;
    var plot = document.getElementById('plot_area');
    var spinner = $('#protein_spinner');

    var tooltips = [];

    function init_plot(variable_config)
    {
        var site_tooltip = Tooltip()

        site_tooltip.init({
            id: 'site',
            template: function(d){
                return nunjucks.render(
                    'site_tooltip.njk',
                    {site: d}
                );
            },
            viewport: plot.parentNode
        })

        tooltips.push(site_tooltip)


        var params = {
            element: plot,
            site_tooltip: site_tooltip,
            sequence_length: {{ protein.sequence | length }},
            paddings: {bottom: 40, top: 30, left: 89, right: 1},
            head_size: tracks.adjustMaxZoom(),
            name: '{{ protein.gene.name }}',
            zoom_callback: tracks.setZoom,
            position_callback: tracks.setAAPosition,
            zoomAndMove_callback: tracks.setZoomAndMove,
            width: tracks_box.width(),
            mutations_color_map: {
                'distal': 'yellow',
                'motif-changing': 'red',
                'network-rewiring': 'red',
                'direct': 'darkred',
                'proximal': 'orange',
                'none': 'darkgray'
            },
            onload: function(){ spinner.hide() }
        };

        update_object(params, variable_config);

        var needle_plot = new NeedlePlot(params);

        return needle_plot;
    }

    $(window).on('resize', function()
        {
            if (tracks.isReady())
                tracks.refreshScaleFactor();
            if(needle_plot)
            {
                needle_plot.setSize(
                    tracks_box.width(),
                    null,
                    tracks.adjustMaxZoom()
                )
            }
        }
    );

    var mutation_table = MutationTable();
    var exporter = Export()

    var $table_wrapper = $('#table-wrapper');

    {% raw %}
    var export_title_pattern = 'protein_view:{{ gene }}-{{ refseq }}-{{ expanded_filters }}';
    {% endraw %}

    {% assets "css_protein" %}
        exporter.init(
            plot,
            'protein_view:{{ protein.gene.name }}-{{ protein.refseq }}-{{ filters.url_string(expanded=True) }}',
            '{{ ASSET_URL }}'
        );
    {% endassets %}

    /**
     * Representation of SequenceView as defined in SequenceView:representation_data().
     * @typedef {RepresentationData} SequenceViewData
     * @property {html} mutation_table
     * @property {html} tracks
     * @property {Object} mutations
     * @property {Object} sites
     * @property {string} value_type
     * @property {boolean} log_scale
     */

    /**
     * @callback for {@see AsyncFiltersHandler.Config.data_handler}
     * Callback  loading representation from server response.
     * @param {SequenceViewData} data
     * @param {FiltersData} filters_data
     */
    function load_data(data, filters_data)
    {
        short_url.reset();

        exporter.update_title(
            format(
                export_title_pattern,
                {
                    gene: '{{ protein.gene.name }}',
                    refseq: '{{ protein.refseq }}',
                    expanded_filters: filters_data.expanded_query
                }
            )
        );

        $table_wrapper.html(data.mutation_table);
        var $table = $('#mutation_table');

        tracks_box.html(data.tracks);
        tracks.init({box: tracks_box.get(0)});

        var max_zoom = tracks.adjustMaxZoom();

        mutation_table.init(
            $table, data.mutations, {{ mutation_types | tojson }}, "{{ url_for('MutationView:show', refseq=protein.refseq, position='{{ pos }}', alt='{{ alt }}') }}"
        );

        var needle_tooltip = Tooltip()
        needle_tooltip.init({
            id: 'needle',
            template: function(mutation) {
                return nunjucks.render(
                    'needle_tooltip.njk',
                    {
                        mutation: mutation,
                        value_type: data.value_type
                    }
                )
            },
            callback: function(node)
            {
                var kinase_tooltip = initializeKinaseTooltips(d3.select(node).selectAll('.kinase'))
                tooltips.push(kinase_tooltip)
            }
        })
        tooltips.push(needle_tooltip)

        var variable_config = {
            needle_tooltip: needle_tooltip,
            legends: {
                x: null,
                y: data.value_type + ' of mutations in {{ protein.gene.name }} {{ protein.refseq }}'
            },
            use_log: data.log_scale,
            data: {
                mutations: data.mutations,
                sites: data.sites
            },
            max_zoom: max_zoom
        };

        needle_plot = init_plot(variable_config);

        tracks.setNeedlePlotInstance(needle_plot);
    }

    var initial_widgets_form = $('#filters_form')
    Widgets.init(initial_widgets_form, function(){});

    var filters_handler = AsyncFiltersHandler();

    filters_handler.init({
        form: initial_widgets_form,
        data_handler: load_data,
        endpoint_url: "{{ url_for('SequenceView:representation_data', refseq=protein.refseq) }}",
        on_loading_start: function (){
            if (tracks.isReady())
                tracks.hide();
            $('#plot_area').addClass('invisible');
            var to_destroy = [needle_plot, tracks, mutation_table]
            var obj, tooltip
            while(obj = to_destroy.pop()) if (obj) obj.destroy();
            while(tooltip = tooltips.pop()) tooltip.remove()
            $table_wrapper.addClass('loading');
            spinner.show();
        },
        on_loading_end: function(){
            $('#plot_area').removeClass('invisible');
            $table_wrapper.removeClass('loading');

            $('.table-help-btn').removeClass('hidden');
            // spinner will be hidden after needle_plot initialization
            Widgets.init($('#filters_form'), function(){});
        },
        links_to_update: $('.variable-url')
    });

    // force on_update event to load initial representation
    filters_handler.on_update(true);

    var state_handlers = {
        filters: filters_handler
    };

    // handle change in history
    $(window).on('popstate', function(event) {
        var state = event.originalEvent.state;
        if(state)
        {
            var handler = state_handlers[state.handler];
            handler.apply(state.filters_query, true, true)
        }
        if(window.location.hash)
        {
            var mutation_id = window.location.hash.substring(1);
            mutation_table.showMutation(mutation_id)
        }
    });

  </script>
  {{ super() }}
{% endblock %}
